microbench/src/main/java/io/netty/handler/codec/http/DecodeHexBenchmark.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.internal.StringUtil;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 1)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+public class DecodeHexBenchmark extends AbstractMicrobenchmark {
+
+    @Param({ "abcdefABCDEF0123456789g", "abcdefABCDEF0123456789" })
+    private String hex;
+    private char[] hexDigits;
+
+    @Setup
+    public void init() {
+        hexDigits = hex.toCharArray();
+    }
+
+    @Benchmark
+    public long hexDigits() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += StringUtil.decodeHexNibble(hexDigits[i]);
+        }
+        return v;
+    }
+
+    @Benchmark
+    public long hexDigitsBranchlessNoChecks() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += branchlessNoCheck(hexDigits[i]);
+        }
+        return v;
+    }
+
+    @Benchmark
+    public long hexDigitsUsingArrayWithChecks() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += lookupWithCheck(hexDigits[i]);
+        }
+        return v;
+    }
+
+    @Benchmark
+    public long hexDigitsUsingArrayWithEx() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += lookupWithEx(hexDigits[i]);
+        }
+        return v;
+    }
+
+    private static final int[] HEX;
+
+    static {
+        HEX = new int['f' + 1];
+        Arrays.fill(HEX, -1);
```

Comment:
Wild idea, make the array `65,535` long, no need for the boundary checks or exception handling. It can be a byte array to save some space (though the locality/waste implications here are pretty minor).

Reply:
I'm going to push a change that uses a byte[] and `PlatformDependent::getByte` to save bounds checks: seems that OpenJDK is duplicating the bounds checks so I'm saving it to do it :D
OpenJ9 seems to work correctly here instead, but I still need to use your advice to randomize more the inputs (that's a nice advice indeed)

Reply:
Changed pushed, but I still want to use relevant bench params for the Netty use case: the original code should behave "quicker" then before when the chars are "right", but shouldn't slow down "too much" when not. I hope this new set of params are better, but I will work on improving the bench later :)